File: packages/sim/src/transcript.ts
// Run transcript = u32 seed (LE) || u8[ticks] buttons.
//
// One file == one run. The seed bootstraps the prng; each subsequent byte is
// the `buttons` value passed to stepMut on that tick. Replay is a pure
// function of (seed, buttons[]) → final GameState — the entire determinism
// guarantee of the sim is staked on this.

import { createInitialState } from "./state.js";
import { Stage } from "./stages.js";
import { stepMut } from "./step.js";
import type { GameState } from "./types.js";

export interface DecodedTranscript {
  seed: number;
  buttons: Uint8Array;
}

export interface ReplayResult {
  state: GameState;
  ticksConsumed: number;  // how many bytes we stepped before gameOver
}

export function encodeTranscript(seed: number, buttons: Uint8Array): Uint8Array {
  const out = new Uint8Array(4 + buttons.length);
  const dv = new DataView(out.buffer);
  dv.setUint32(0, seed >>> 0, true); // little-endian — matches Rust default
  out.set(buttons, 4);
  return out;
}

export function decodeTranscript(buf: Uint8Array): DecodedTranscript {
  if (buf.length < 4) throw new Error(`transcript too short (${buf.length} bytes)`);
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const seed = dv.getUint32(0, true);
  return { seed, buttons: buf.slice(4) };
}

/** Replay a transcript end-to-end. Stops as soon as state.gameOver flips, since
 *  subsequent steps are no-ops; returns the tick count we actually consumed. */
export function replay(buf: Uint8Array, startStage: Stage = Stage.Common): ReplayResult {
  const { seed, buttons } = decodeTranscript(buf);
  const state = createInitialState(seed, startStage);
  let i = 0;
  for (; i < buttons.length; i++) {
    stepMut(state, { buttons: buttons[i]! });
    if (state.gameOver) {
      i++;
      break;
    }
  }
  return { state, ticksConsumed: i };
}

File: services/prover/core/src/transcript.rs
//! Transcript codec + replay — mirror of `packages/sim/src/transcript.ts`.
//!
//! Layout matches TS: `u32 seed (LE) || u8[ticks] buttons`. Same `replay`
//! semantics: step until either we exhaust buttons or `game_over` flips.

use crate::stages::Stage;
use crate::state::create_initial_state;
use crate::step::step_mut;
use crate::types::{GameState, PlayerInput};

pub struct DecodedTranscript {
    pub seed: i32,
    pub buttons: Vec<u8>,
}

pub struct ReplayResult {
    pub state: GameState,
    pub ticks_consumed: u32,
}

pub fn encode_transcript(seed: i32, buttons: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(4 + buttons.len());
    out.extend_from_slice(&(seed as u32).to_le_bytes());
    out.extend_from_slice(buttons);
    out
}

#[derive(Debug)]
pub enum TranscriptError {
    TooShort(usize),
}

impl std::fmt::Display for TranscriptError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TranscriptError::TooShort(n) => write!(f, "transcript too short ({n} bytes)"),
        }
    }
}

impl std::error::Error for TranscriptError {}

pub fn decode_transcript(buf: &[u8]) -> Result<DecodedTranscript, TranscriptError> {
    if buf.len() < 4 {
        return Err(TranscriptError::TooShort(buf.len()));
    }
    let seed = u32::from_le_bytes([buf[0], buf[1], buf[2], buf[3]]) as i32;
    let buttons = buf[4..].to_vec();
    Ok(DecodedTranscript { seed, buttons })
}

/// Replay a transcript end-to-end. Stops as soon as `game_over` flips, since
/// further steps are no-ops; returns the number of ticks consumed.
pub fn replay(buf: &[u8], start_stage: Stage) -> Result<ReplayResult, TranscriptError> {
    let DecodedTranscript { seed, buttons } = decode_transcript(buf)?;
    let mut state = create_initial_state(seed, start_stage);
    let mut consumed: u32 = 0;
    for &b in &buttons {
        step_mut(&mut state, PlayerInput { buttons: b });
        consumed += 1;
        if state.game_over {
            break;
        }
    }
    Ok(ReplayResult { state, ticks_consumed: consumed })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_then_decode_round_trips() {
        let seed = 0x5e570e03_i32;
        let buttons = vec![0u8, 1, 2, 8, 0, 0, 0, 9];
        let buf = encode_transcript(seed, &buttons);
        assert_eq!(buf.len(), 4 + buttons.len());
        let decoded = decode_transcript(&buf).unwrap();
        assert_eq!(decoded.seed, seed);
        assert_eq!(decoded.buttons, buttons);
    }

    #[test]
    fn decode_rejects_short_input() {
        assert!(matches!(
            decode_transcript(&[0u8, 1, 2]),
            Err(TranscriptError::TooShort(3))
        ));
    }

    #[test]
    fn empty_replay_does_not_step() {
        let buf = encode_transcript(123, &[]);
        let r = replay(&buf, Stage::Common).unwrap();
        assert_eq!(r.ticks_consumed, 0);
        assert_eq!(r.state.tick, 0);
        assert!(!r.state.game_over);
    }

    #[test]
    fn replay_stops_on_game_over() {
        // 200 ticks of UP at fp(6)/tick should hit WorldTop well inside the
        // buffer; replay must stop early.
        let buf = encode_transcript(1, &vec![crate::types::BTN_UP; 200]);
        let r = replay(&buf, Stage::Common).unwrap();
        assert!(r.state.game_over);
        assert!(r.ticks_consumed < 200);
        assert_eq!(r.state.game_over_reason, crate::types::GameOverReason::WorldTop);
    }
}
